src/Infrastructure/Service/Precondition/AbstractLinkIteratingPrecondition.php

```diff
@@ -0,0 +1,122 @@
+
+    protected PathFactoryInterface $pathFactory;
+
+    final protected function getUnfulfilledStatusMessage(): string
+    {
+        return $this->defaultUnfulfilledStatusMessage;
+    }
+
+    abstract protected function getDefaultUnfulfilledStatusMessage(): string;
+
+    /** @throws \PhpTuf\ComposerStager\Domain\Exception\IOException */
+    abstract protected function isSupportedLink(PathInterface $file, PathInterface $codebaseRootDir): bool;
+
+    public function __construct(
+        RecursiveFileFinderInterface $fileFinder,
+        FilesystemInterface $filesystem,
+        PathFactoryInterface $pathFactory
+    ) {
+        $this->fileFinder = $fileFinder;
+        $this->filesystem = $filesystem;
+        $this->pathFactory = $pathFactory;
+        $this->defaultUnfulfilledStatusMessage = $this->getDefaultUnfulfilledStatusMessage();
+    }
+
+    public function isFulfilled(
+        PathInterface $activeDir,
+        PathInterface $stagingDir,
+        ?PathListInterface $exclusions = null
+    ): bool {
+        try {
+            $exclusions ??= new PathList([]);
+            $exclusions->add([$stagingDir->resolve()]);
+
+            if ($this->exitEarly($activeDir, $stagingDir, $exclusions)) {
+                return true;
+            }
+
+            $directories = [
+                'active' => $activeDir,
+                'staging' => $stagingDir,
+            ];
+
+            foreach ($directories as $name => $directoryRoot) {
+                $files = $this->findFiles($directoryRoot, $exclusions);
+
+                foreach ($files as $file) {
+                    $file = $this->pathFactory::create($file);
+
+                    if (!$this->filesystem->isLink($file)) {
+                        continue;
+                    }
+
+                    if (!$this->isSupportedLink($file, $directoryRoot)) {
+                        $this->defaultUnfulfilledStatusMessage = sprintf(
+                            $this->defaultUnfulfilledStatusMessage,
+                            $name,
+                            $directoryRoot->resolve(),
+                            $file->resolve(),
+                        );
+
+                        return false;
+                    }
+                }
+            }
+        } catch (InvalidArgumentException|IOException $e) {
+            // If something goes wrong, don't throw an exception--just consider the precondition
+            // unfulfilled and pass details along to the user via the status message.
+            // @todo Find a way to bubble this exception up as the $previous
+            //   argument to PreconditionException in ::assertIsFulfilled().
+            $this->defaultUnfulfilledStatusMessage = $e->getMessage();
+
+            return false;
+        }
+
+        return true;
+    }
+
+    /** Determines whether to exit the "is fulfilled" test early, before expensive scanning for links. */
+    protected function exitEarly(
```

Comment:
I don't object to this, but I'm curious, what are some cases where this would be useful for a subclass to override?